### PR TITLE
Record Table

### PR DESCRIPTION
**Pull request by darth-tytus**
- Introduce record table
- use it to replace InterpreterRecords and CompiledRecord (unify record interface)

This pull request is a work in progress.

### CODE REVIEW

File: src/RamOperation.h [hunk 1]
```diff
@@ -510,7 +510,7 @@ class RamParallelChoice : public RamChoice, public RamAbstractParallel {
     RamParallelChoice(std::unique_ptr<RamRelationReference> rel, size_t ident,
             std::unique_ptr<RamCondition> cond, std::unique_ptr<RamOperation> nested,
             std::string profileText = "")
-            : RamChoice(std::move(rel), ident, std::move(cond), std::move(nested), profileText) {}
+            : RamChoice(std::move(rel), ident, std::move(cond), std::move(nested), std::move(profileText)) {}
```
Comment (b-scholz) [inline src/RamOperation.h]: Please revert this change; it does not affect the record table.

File: src/RecordTable.h [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+
+/**
+ * A bidirectional mapping between tuples and reference indices.
+ */
+class RecordMap {
+    /** The arity of the stored tuples */
+    int arity;
+
+    /** The mapping from tuples to references/indices */
+    std::map<std::vector<RamDomain>, RamDomain> r2i;
```
Comment (b-scholz) [inline src/RecordTable.h]: Please, use a better name for this map.

File: src/RecordTable.h [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+
+    /** The mapping from indices to tuples */
+    std::vector<std::vector<RamDomain>> i2r;
```
Comment (b-scholz) [inline src/RecordTable.h]: Please, use a better name for this vector.

File: src/RecordTable.h [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+            tuple.data[i] = data[i];
+        }
+        return tuple;
+    }
+
+    /**
+     * Determines whether the given reference is the null reference encoding
+     * the absence of any nested record.
+     */
+    bool isNull(RamDomain ref) {
```
Comment (b-scholz) [inline src/RecordTable.h]: This should be renamed to isNil()  -- we use "nil" is a syntactic element to terminate a record.

File: src/RecordTable.h [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+        return ref == 0;
+    }
+
+private:
+    mutable Lock access;
```
Comment (b-scholz) [inline src/RecordTable.h]: can we try an OMP-directive instead of a Lock just for keeping it consistent?

File: src/Synthesiser.cpp [hunk 6]
```diff
@@ -819,8 +815,15 @@ void Synthesiser::emitCode(std::ostream& out, const RamStatement& stmt) {
             visit(lookup.getExpression(), out);
             out << ";\n";
 
-            out << "if (isNull<" << tuple_type << ">(ref)) continue;\n";
-            out << tuple_type << " env" << lookup.getTupleId() << " = unpack<" << tuple_type << ">(ref);\n";
+            // Handle null case.
```
Comment (b-scholz) [inline src/Synthesiser.cpp]: // Handle nil case.

File: src/Makefile.am [hunk 7]
```diff
@@ -180,7 +180,6 @@ soufflepublic_HEADERS = \
         Brie.h                                    \
         BTree.h                                   \
         CompiledIndexUtils.h                      \
-        CompiledRecord.h                          \
```
Comment (mmcgr) [inline src/Makefile.am]: RecordTable.h is needed in the public headers, as the replacement for CompiledRecord.h

File: src/RecordTable.h [hunk 8]
```diff
@@ -0,0 +1,158 @@
...
+     * A function obtaining a pointer to the tuple addressed by the given reference.
+     */
+    RamDomain* unpack(RamDomain ref, int arity) {
+        return getForArity(arity).unpack(ref);
+    }
+
+    /**
+     * A function obtaining a pointer to the tuple addressed by the given reference.
+     */
+    template <typename Domain, std::size_t _arity>
```
Comment (mmcgr) [inline src/RecordTable.h]: You could use `Arity` instead of `_arity`, which would be more in keeping with our usual naming strategies.

File: src/Makefile.am (hunk 7 above)
Comment (b-scholz) [reply src/Makefile.am]: Also please check the file ```configure.ac``` whether RecordTable.h is listed there.
